src/garage/tf/algos/npo.py

```diff
@@ -474,53 +442,64 @@ def _build_entropy_term(self, i):
+                optimization input values.
+            observations (np.ndarray): Padded observations.
+            valids (np.ndarrsy): 1-D array indicating valid values of
+                observations.
 
-        """
-        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+        Returns:
+            np.ndarray: Padded augment returns.
 
-        # Augment reward from baselines
-        rewards_tensor = self._f_rewards(*policy_opt_input_values)
+        """
         returns_tensor = self._f_returns(*policy_opt_input_values)
         returns_tensor = np.squeeze(returns_tensor, -1)
 
-        paths = samples_data['paths']
-        valids = samples_data['valids']
-
-        # Recompute parts of samples_data
-        aug_rewards = []
+        paths = []
         aug_returns = []
-        for rew, ret, val, path in zip(rewards_tensor, returns_tensor, valids,
-                                       paths):
-            path['rewards'] = rew[val.astype(np.bool)]
-            path['returns'] = ret[val.astype(np.bool)]
-            aug_rewards.append(path['rewards'])
-            aug_returns.append(path['returns'])
-        samples_data['rewards'] = pad_tensor_n(aug_rewards,
-                                               self.max_episode_length)
-        samples_data['returns'] = pad_tensor_n(aug_returns,
-                                               self.max_episode_length)
+
+        for ret, val, ob in zip(returns_tensor, valids, observations):
+            returns = ret[val.astype(np.bool)]
+            obs = ob[val.astype(np.bool)]
+            paths.append(dict(observations=obs, returns=returns))
+            aug_returns.append(returns)
+
+        aug_returns = pad_tensor_n(aug_returns, self.max_episode_length)
 
         # Fit baseline
         logger.log('Fitting baseline...')
         self._baseline.fit(paths)
+        return aug_returns
 
-    def _policy_opt_input_values(self, samples_data):
-        """Map episode samples to the policy optimizer inputs.
+    def _process_episodes_batch(self, episodes):
+        """Process batch of episodes to the policy optimizer inputs.
 
         Args:
-            samples_data (dict): Processed sample data.
-                See garage.tf.paths_to_tensors() for details.
+            episodes (EpisodeBatch): Batch of episodes.
 
         Returns:
             list(np.ndarray): Flatten policy optimization input values.
+            np.ndarray: Padded observations.
+            np.ndarray: Padded baseline predictions.
+            np.ndarrsy: 1-D array indicating valid values of observations and
+                baselines.
 
         """
+        paths = episodes.to_list()
+        for i, _ in enumerate(paths):
+            paths[i]['actions'] = self._env_spec.action_space.flatten_n(
+                paths[i]['actions'])
+
+        baselines = [self._baseline.predict(path) for path in paths]
+        samples_data = paths_to_tensors(paths, self.max_episode_length,
```

Comment:
is there any way we can pass EpisodeBatch + a few other tensors to TensorFlow? A good detector that #1110 is done is "we can delete all usages of `paths_to_tensors`"

Reply:
I think if padding the tensor is necessary, we must split the EpisodeBatch.

Reply:
Or we can do the padding operation in TensorFlow.

Reply:
@ryanjulian Can we do the padding operation in EpisodeBatch. For example, `episode_batch.padded_list()`. I have tried to do all things in tensorflow, but it is difficult as the padding size is a variable.